src/coreclr/jit/flowgraph.cpp

```diff
@@ -436,6 +436,268 @@ BasicBlock* Compiler::fgCreateGCPoll(GCPollType pollType, BasicBlock* block)
+                int            isInitOffset;
+                InfoAccessType staticBaseAccessType;
+                size_t         staticBase;
+                size_t         staticInitAddr =
+                    info.compCompHnd->getIsClassInitedFieldAddress(call->gtRetClsHnd, isGc, &staticBaseAccessType,
+                                                                   &staticBase, &isInitMask, &isInitOffset);
+                if (staticInitAddr == 0)
+                {
+                    JITDUMP("getIsClassInitedFieldAddress returned 0 - bail out.\n")
+                    continue;
+                }
+                assert((staticBase != 0) && (isInitMask != 0));
+
+                DebugInfo debugInfo = stmt->GetDebugInfo();
+
+                // Split block right before the call tree
+                BasicBlock* prevBb       = block;
+                GenTree**   callUse      = nullptr;
+                Statement*  newFirstStmt = nullptr;
+                block                    = fgSplitBlockBeforeTree(block, stmt, call, &newFirstStmt, &callUse);
+                assert(prevBb != nullptr && block != nullptr);
+
+                // Block ops inserted by the split need to be morphed here since we are after morph.
+                // We cannot morph stmt yet as we may modify it further below, and the morphing
+                // could invalidate callUse.
+                while ((newFirstStmt != nullptr) && (newFirstStmt != stmt))
+                {
+                    fgMorphStmtBlockOps(block, newFirstStmt);
+                    newFirstStmt = newFirstStmt->GetNextStmt();
+                }
+
+                GenTree* staticBaseTree;
+                if (staticBaseAccessType == IAT_VALUE)
+                {
+                    staticBaseTree = gtNewIconHandleNode(staticBase, GTF_ICON_STATIC_HDL);
+                }
+                else
+                {
+                    assert(staticBaseAccessType == IAT_PVALUE);
+                    staticBaseTree = gtNewIndOfIconHandleNode(TYP_I_IMPL, staticBase, GTF_ICON_STATIC_ADDR_PTR, true);
+                }
+
+                *callUse = staticBaseTree;
+
+                fgMorphStmtBlockOps(block, stmt);
+                gtUpdateStmtSideEffects(stmt);
+
+                //
+                // Create new blocks. Essentially, we want to transform this:
+                //
+                //   staticBase = helperCall();
+                //
+                // into:
+                //
+                //   if (!isInitialized)
+                //   {
+                //       helperCall(); // we don't use its return value
+                //   }
+                //   staticBase = fastPath;
+                //
+
+                // The initialization check looks like this for JIT:
+                //
+                // *  JTRUE     void
+                // \--*  NE        int
+                //    +--*  AND       int
+                //    |  +--*  IND       int
```

Comment:
The flag is just a `byte` for standard CoreCLR.

I think it is fine to hardcode the byte vs. int different in the JIT based on `IsTargetAbi(CORINFO_NATIVEAOT_ABI)`. It is not very valuable to try to abstract it through JIT/EE interface.

Reply:
We can flip the logic for native AOT to make 0 means initialized. It should make the code for arm64 smaller thanks for cbz/cbnz. Another reason that abstracting this detail via JIT/EE interface is hard.

Reply:
> We can flip the logic for native AOT to make 0 means initialized.

Draft: https://github.com/dotnet/runtime/pull/83937

Reply:
Here is the current arm64 codegen for NativeAOT win-arm64:
```asm
; Method GetInt():int:this
        A9BF7BFD          stp     fp, lr, [sp, #-0x10]!
        910003FD          mov     fp, sp
        90000000          adrp    x0, [HIGH RELOC #0x40000000004216f8]
        91000000          add     x0, x0, [LOW RELOC #0x40000000004216f8]
        B85F8000          ldr     w0, [x0, #-0x08]
        7100041F          cmp     w0, #1
        540000C1          bne     G_M24736_IG05
G_M24736_IG03:              
        90000000          adrp    x0, [HIGH RELOC #0x40000000004216f8]
        91000000          add     x0, x0, [LOW RELOC #0x40000000004216f8]
        B9400000          ldr     w0, [x0]
        A8C17BFD          ldp     fp, lr, [sp], #0x10
        D65F03C0          ret     lr
G_M24736_IG05:              
        94000000          bl      CORINFO_HELP_READYTORUN_NONGCSTATIC_BASE
        17FFFFFA          b       G_M24736_IG03
; Total bytes of code: 56
```
relocs look terrible 🙁.

I guess the idea to replace
```asm
        7100041F          cmp     w0, #1
        540000C1          bne     G_M24736_IG05
```
with
```asm
        540000C1          tbnz    w0, G_M24736_IG05
```